manager/controllers/app/npIsolation.go

```diff
@@ -0,0 +1,344 @@
+				from = append(from, npPeer)
+			} else {
+				log.Warn().Str(logging.ACTION, logging.CREATE).Msgf("Ingress has empty release %v", ingress)
+			}
+		} else {
+			// TODO: multi-cluster support
+			log.Debug().Str(logging.ACTION, logging.CREATE).Msgf("Cross-cluster ingress connectivity, ingress.Cluster %s, blueprint cluster %s",
+				ingress.Cluster, cluster)
+		}
+	}
+	if len(from) == 0 {
+		return getDefaultNetworkPolicyIngressRules(), nil
+	}
+	return []netv1.NetworkPolicyIngressRule{{From: from, Ports: []netv1.NetworkPolicyPort{{Protocol: &tcp}}}}, nil
+}
+
+func (r *BlueprintReconciler) createNPEgressRules(ctx context.Context, egresses []fapp.ModuleDeployment, urls []string,
+	cluster string, modulesNamespace string, log *zerolog.Logger) []netv1.NetworkPolicyEgressRule {
+	log.Trace().Str(logging.ACTION, logging.CREATE).
+		Msgf("Egress rules creation from network.Egress: %v, network.URLs %v",
+			egresses, urls)
+	egressRules := []netv1.NetworkPolicyEgressRule{}
+
+	modulesEgressRules := r.createNextModulesEgressRules(egresses, cluster, log)
+	egressRules = append(egressRules, modulesEgressRules...)
+
+	// The URLs can be a CIDR block, or urls to the cluster internal or external services.
+	for _, urlString := range urls {
+		log.Trace().Msgf("Processing external URL %s", urlString)
+
+		// Check if it is a CIDR (Classless Inter-Domain Routing)
+		_, _, err := net.ParseCIDR(urlString)
+		if err == nil {
+			ipBlock := netv1.IPBlock{CIDR: urlString}
+			to := []netv1.NetworkPolicyPeer{{IPBlock: &ipBlock}}
+			egressRules = append(egressRules, netv1.NetworkPolicyEgressRule{To: to})
+			continue
+		}
+		// parse URL
+		servURL, err := managerUtils.ParseRawURL(urlString)
+		if err != nil {
+			log.Err(err).Msgf(CannotParseURLError, urlString)
+			continue
+		}
+		hostName := servURL.Hostname()
+		if hostName == "" {
+			log.Warn().Msgf("URL without host name: %s", servURL)
+			continue
+		}
+		// Check if the hostName is actually an IP address.
+		// NOTE: IP address to a local service will not work
+		ip := net.ParseIP(hostName)
+		if ip != nil {
+			ipBlock := ipToIPBlock(ip)
+			to := []netv1.NetworkPolicyPeer{{IPBlock: &ipBlock}}
+			policyPort := policyPortFromURL(servURL, log)
+			egressRules = append(egressRules, netv1.NetworkPolicyEgressRule{To: to, Ports: []netv1.NetworkPolicyPort{policyPort}})
+			continue
+		}
+		// Check if it is a local service
+		hostStrings := strings.Split(hostName, ".")
+		// local K8s services have the following format: <serviceName>.<nameSpace>.svc.cluster.local
+		if len(hostStrings) < 3 || hostStrings[2] == "svc" { //nolint: revive
+			service := corev1.Service{}
+			key := types.NamespacedName{Name: hostStrings[0]}
+			if len(hostStrings) > 1 {
+				key.Namespace = hostStrings[1]
+			} else {
+				key.Namespace = modulesNamespace
+			}
+			// we assume that the service exists
+			if err = r.Get(ctx, key, &service); err != nil {
+				log.Info().Msgf("Get service returned error. %v", service)
```

Comment:
Warning level?

Reply:
If you don't have a ClusterRole for accessing services, and running in namespace-scope, the Get command will fail, I think. There are no cache selectors for Service in manager/main.go.

Reply:
I thought about it, there are several points here:

1. I cannot separate external services, e.g. `docker.io` from internal services, e.g. `vault.fybrik-system`. If we can assume/request that all internal services will be defined with at least the `svc` suffix, e.g. `vault.fybrik-system.svc` or with its FQDN - `vault.fybrik-system.svc.cluster.local`. Without this separation, failures will happen anyway. I want to return to this point when @Mohammad-nassar10 finishes his multi-cluster work. 

2. A cache selector for Services is missing by purpose. I don't want to cache Service object because they are not requested by labels or some other attributes, only by key. Furthermore, I added Services into `ClientDisableCacheFor` manager attribute (main.go:110), without it if we don't have the `ClusterScope` permissions, the cache reflector fails with the following message:

> {"level":"error","time":"2023-04-25T06:36:56Z","caller":"/home/roytman/go/pkg/mod/k8s.io/client-go@v0.25.2/tools/cache/reflector.go:140","message":"pkg/mod/k8s.io/client-go@v0.25.2/tools/cache/reflector.go:169: Failed to watch *v1.Service: failed to list *v1.Service: services is forbidden: User \"system:serviceaccount:fybrik-system:manager\" cannot list resource \"services\" in API group \"\" at the cluster scope\n"}

3. The error message is printed in the Info level, because again I cannot separate local services vs external, and don't want to "afraid" potential users. If we resolve the first bullet, we can change the level to Warn or even Error here.